### PR TITLE
cmake: errors when using SDL_sound as submodule

### PR DESCRIPTION
**Pull request by ccvelandres**
Genexp doesn't work in cmake versions below 3.26 in certain usage. (latest cmake release is still 3.25)

https://github.com/icculus/SDL_sound/blob/6cb07c21fab468be5f6a8b4c00c1b29697a9cc7a/CMakeLists.txt#L248

See note in https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_PROPERTY for reference.

Directory looks like this:
```bash
├── library1
│   ├── CMakeLists.txt              <- library_1 links to SDL_Sound
│   └── SDL_Sound                   <- submodule
├── example
│   └── CMakeLists.txt              <- exe links to library_1 
└── CMakeLists.txt                   <- root CMakeLists
```

```bash
CMake Error at example/CMakeLists.txt:14 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_PROPERTY:SDL2::SDL2,INTERFACE_INCLUDE_DIRECTORIES>

  Target "SDL2::SDL2" not found.
```

### CODE REVIEW

**Comment by ccvelandres** [comment]
53df0acd4cc6688e1b8584a56e6befd2637415d7 -- Adds include directories when using SDL_sound as submodule

**Comment by sezero** [comment]
CC: @madebr

File: CMakeLists.txt [hunk 1]
```diff
@@ -212,10 +212,12 @@ if(SDLSOUND_BUILD_STATIC)
     if(NOT MSVC)
         set_target_properties(SDL2_sound-static PROPERTIES OUTPUT_NAME "SDL2_sound")
     endif()
-    target_include_directories(SDL2_sound-static PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>)
+    target_include_directories(SDL2_sound-static PUBLIC 
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
```
Comment (madebr) [inline CMakeLists.txt:216]: Let's also allow `#include "SDL2/SDL_sound.h"` for installed prefix'es.
This method of inclusion is currently also supported by SDL3.

```suggestion
        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -212,10 +212,12 @@ if(SDLSOUND_BUILD_STATIC)
...
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
     if(TARGET SDL2::SDL2)
-        target_include_directories(SDL2_sound-static
-            PUBLIC $<TARGET_PROPERTY:SDL2::SDL2,INTERFACE_INCLUDE_DIRECTORIES>)
+        get_target_property(_sdl2_include_dir SDL2::SDL2 INTERFACE_INCLUDE_DIRECTORIES)
+        target_include_directories(SDL2_sound-static PUBLIC "${_sdl2_include_dir}")
         target_link_libraries(SDL2_sound-static PRIVATE $<BUILD_INTERFACE:SDL2::SDL2> ${OPTIONAL_LIBRARY_LIBS} ${OTHER_LDFLAGS})
     else()
         target_include_directories(SDL2_sound-static PUBLIC ${SDL2_INCLUDE_DIRS})
```
Comment (madebr) [inline CMakeLists.txt:223]:
```suggestion
        target_include_directories(SDL2_sound-static PUBLIC $<BUILD_INTERFACE:${SDL2_INCLUDE_DIRS}>)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -241,11 +243,13 @@ if(SDLSOUND_BUILD_SHARED)
     if(OS2)   # OS/2 does not support a DLL name longer than 8 characters.
         SET_TARGET_PROPERTIES(SDL2_sound PROPERTIES OUTPUT_NAME "SDL2snd")
     endif()
-    target_include_directories(SDL2_sound
-        PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
+    target_include_directories(SDL2_sound PUBLIC 
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
```
Comment (madebr) [inline CMakeLists.txt:247]: Allow `#include "SDL2/SDL_sound.h"` as well.
```suggestion
        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -212,10 +212,12 @@ if(SDLSOUND_BUILD_STATIC)
...
     endif()
-    target_include_directories(SDL2_sound-static PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>)
+    target_include_directories(SDL2_sound-static PUBLIC 
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
     if(TARGET SDL2::SDL2)
-        target_include_directories(SDL2_sound-static
-            PUBLIC $<TARGET_PROPERTY:SDL2::SDL2,INTERFACE_INCLUDE_DIRECTORIES>)
+        get_target_property(_sdl2_include_dir SDL2::SDL2 INTERFACE_INCLUDE_DIRECTORIES)
+        target_include_directories(SDL2_sound-static PUBLIC "${_sdl2_include_dir}")
```
Comment (madebr) [inline CMakeLists.txt:220]: We want to keep these paths hidden from the installed cmake config files.
```suggestion
        get_target_property(_sdl2_include_dir SDL2::SDL2 INTERFACE_INCLUDE_DIRECTORIES)
        target_include_directories(SDL2_sound-static PUBLIC $<BUILD_INTERFACE:${_sdl2_include_dir}>)
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -241,11 +243,13 @@ if(SDLSOUND_BUILD_SHARED)
...
-    target_include_directories(SDL2_sound
-        PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
+    target_include_directories(SDL2_sound PUBLIC 
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
     )
     if(TARGET SDL2::SDL2)
-        target_include_directories(SDL2_sound PUBLIC "$<TARGET_PROPERTY:SDL2::SDL2,INTERFACE_INCLUDE_DIRECTORIES>")
+        get_target_property(_sdl2_include_dir SDL2::SDL2 INTERFACE_INCLUDE_DIRECTORIES)
+        target_include_directories(SDL2_sound PUBLIC "${_sdl2_include_dir}")
```
Comment (madebr) [inline CMakeLists.txt:252]:
```suggestion
        get_target_property(_sdl2_include_dir SDL2::SDL2 INTERFACE_INCLUDE_DIRECTORIES)
        target_include_directories(SDL2_sound PUBLIC $<BUILD_INTERFACE:${_sdl2_include_dir}>)
```

File: CMakeLists.txt (hunk 1 above)
Comment (ccvelandres) [reply CMakeLists.txt:217]: Done with f0b29fc0ea8c96da4198bf7adf75279acf43ad2c

File: CMakeLists.txt (hunk 3 above)
Comment (ccvelandres) [reply CMakeLists.txt:249]: Done with f0b29fc0ea8c96da4198bf7adf75279acf43ad2c

File: CMakeLists.txt (hunk 2 above)
Comment (ccvelandres) [reply CMakeLists.txt]: Done with 83df0f87e9052f32b30c32dba7128a2384296c5c

File: CMakeLists.txt (hunk 5 above)
Comment (ccvelandres) [reply CMakeLists.txt]: Done with 83df0f87e9052f32b30c32dba7128a2384296c5c

File: CMakeLists.txt (hunk 4 above)
Comment (ccvelandres) [reply CMakeLists.txt]: Done with 83df0f87e9052f32b30c32dba7128a2384296c5c

**Comment by sezero** [comment]
@madebr: Is this good now?

**Comment by madebr** [comment]
https://github.com/ccvelandres/SDL_sound/pull/1 contains my (locally tested) suggestions

**Comment by ccvelandres** [comment]
Merged https://github.com/ccvelandres/SDL_sound/pull/1 and also tested locally on my end

**Comment by ccvelandres** [comment]
One other thing I've noticed is this: we are disabling RPATH for SDL2_sound which causes this error: `cannot open shared object file: No such file or directory`. 

In SDL, RPATH depends on ${UNIX_SYS} --> https://github.com/libsdl-org/SDL/blob/SDL2/CMakeLists.txt#L165

Should we enable RPATH for unix systems same as SDL?

File: CMakeLists.txt [hunk 6]
```diff
@@ -205,28 +205,33 @@ if(OS2) # no static library on os/2
...
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
+    if(TARGET SDL2::SDL2-static)
+        get_target_property(_sdl2_include_dir SDL2::SDL2-static INTERFACE_INCLUDE_DIRECTORIES)
+        target_include_directories(SDL2_sound-static PUBLIC "$<BUILD_INTERFACE:${_sdl2_include_dir}>")
+        target_link_libraries(SDL2_sound-static PRIVATE $<BUILD_INTERFACE:SDL2::SDL2-static> ${OPTIONAL_LIBRARY_LIBS} ${OTHER_LDFLAGS})
     else()
-        target_include_directories(SDL2_sound-static PUBLIC ${SDL2_INCLUDE_DIRS})
-        target_link_libraries(SDL2_sound-static PRIVATE $<BUILD_INTERFACE:${SDL2_LIBRARIES}> ${OPTIONAL_LIBRARY_LIBS} ${OTHER_LDFLAGS})
+        target_include_directories(SDL2_sound-static PUBLIC $<BUILD_INTERFACE:${SDL2_INCLUDE_DIRS}>)
```
Comment (madebr) [inline CMakeLists.txt:225]: Without these, build system paths were exported in `SDL2_sound-static-targets.cmake`.
```suggestion
        target_include_directories(SDL2_sound-static PUBLIC "$<BUILD_INTERFACE:${SDL2_INCLUDE_DIRS}>")
```

**Comment by madebr** [comment]
> Should we enable RPATH for unix systems same as SDL? or enable BUILD_RPATH and skip INSTALL_RPATH?

Personally, I would remove [these 2 lines](https://github.com/icculus/SDL_sound/blob/6cb07c21fab468be5f6a8b4c00c1b29697a9cc7a/CMakeLists.txt#L24-L25).
The cmake scripts of SDL_mixer/SDL_image/SDL_ttf don't care about the runpath.

The default behavior of CMake is imho fine: use a RUNPATH in the build directory, but remove it when installing.

File: CMakeLists.txt (hunk 6 above)
Comment (ccvelandres) [reply CMakeLists.txt]: Fixed on https://github.com/icculus/SDL_sound/pull/81/commits/c9851e15eea3fcd749541c7fb87ed04d50f5b00e

**Comment by sezero** [comment]
Thanks.

**Pull request merged**